### PR TITLE
[JENKINS-71639] Allow injecting a securityContext for jnlp containers when using restricted PSS

### PR DESCRIPTION
**Pull request by aneveux**
Hello folks :wave:

Opening this PR to propose an implementation addressing https://issues.jenkins.io/browse/JENKINS-71639.

image

With this new setting, it is possible to automatically inject a `securityContext` configuration in `jnlp` containers.

As detailed in the Jira issue, the configuration which will be used is the following:

```
    securityContext: 
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      seccompProfile:
        type: RuntimeDefault 
```

I added an automated test to ensure this configuration would be injected when the configuration is set.

Let me know if I missed anything.

Thanks a lot for the reviews!

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 1]
```diff
@@ -302,10 +304,29 @@ public Pod build() {
...
         // default jnlp container
-        Optional<Container> jnlpOpt = pod.getSpec().getContainers().stream().filter(c -> JNLP_NAME.equals(c.getName()))
-                .findFirst();
-        Container jnlp = jnlpOpt.orElse(new ContainerBuilder().withName(JNLP_NAME)
-                .withVolumeMounts(volumeMounts.values().toArray(new VolumeMount[volumeMounts.values().size()])).build());
+        Optional<Container> jnlpOpt = pod.getSpec().getContainers().stream().filter(c -> JNLP_NAME.equals(c.getName())).findFirst();
+        Container jnlp = jnlpOpt.orElse(new ContainerBuilder().withName(JNLP_NAME).withVolumeMounts(
+                volumeMounts.values().toArray(new VolumeMount[volumeMounts.values().size()])).build());
+
+        // JENKINS-71639 - Inject `securityContext` in `jnlp` container definition when option activated
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:311]: I think you want to use https://github.com/jenkinsci/kubernetes-plugin/blob/746e28c7546b7156692a162927cd52129762d8fb/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pod/decorator/PodDecorator.java#L10-L12 rather than hard-coding this here. In fact this sort of thing could be done by a separate plugin.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly [hunk 2]
```diff
@@ -31,6 +31,10 @@
         <f:textbox/>
       </f:entry>
 
+      <f:entry title="${%Inject restricted PSP security context in JNLP container definition}" field="pspSecurityContext">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly:34]: Again, “PSP”?

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-pspSecurityContext.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    <p>
+        This option allows to automatically inject in the <code>jnlp</code> container a <code>securityContext</code> that is suitable for the use
+        of the <code>restricted</code> <a href="https://kubernetes.io/docs/concepts/security/pod-security-standards/">Pod Security Standard</a>.
+    </p>
+    <p>
+        This is the <code>securityContext</code> that will be used for the <code>jnlp</code> container with that option:
+    </p>
+    <pre>
+        <code>
+            securityContext:
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-pspSecurityContext.html:16]: Check rendering; I suspect this would be indented much too much in the GUI.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 4]
```diff
@@ -158,6 +159,28 @@ public void testBuildJnlpFromYamlWithNullEnv() throws Exception {
...
+        cloud.setPspSecurityContext(true);
+        PodTemplate template = new PodTemplate();
+        template.setYaml(loadYamlFile("pod-busybox.yaml"));
+        setupStubs();
+        Pod pod = new PodTemplateBuilder(template, slave).build();
+        Map<String, Container> containers = toContainerMap(pod);
+        assertTrue(containers.containsKey("jnlp"));
+
+        Container jnlp = containers.get("jnlp");
+        assertNotNull(jnlp.getSecurityContext());
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:174]: Fine enough, but more interesting would be a Kind-based test that actually enables PSA and verifies that you can run a build using this mode.

**Review by jeromepochat — CHANGES_REQUESTED** [review]
I guess you mean PSS (Pod Security Standards) instead of PSP in names?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 5]
```diff
@@ -112,6 +112,7 @@ public class KubernetesCloud extends Cloud {
 
     private String namespace;
     private String jnlpregistry;
+    private boolean pspSecurityContext = false;
```
[the remark this answers is not in the document]
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: :facepalm: Yes, definitely did not mean `psp`. Too many acronyms :stuck_out_tongue:

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly (hunk 2 above)
Comment (aneveux) [reply src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly]: :facepalm: Changed it, my mistake.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-pspSecurityContext.html (hunk 3 above)
Comment (aneveux) [reply src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-pspSecurityContext.html]: Rendering is illustrated in the PR description, it was indeed probably a bit too much indented. I updated to:

image

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 1 above)
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Oh that's indeed what I wanted to use! Thanks a lot for the pointer :heart: 

I refactored the code to use it.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java (hunk 4 above)
Comment (aneveux) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:174]: I'll give it a try and see if I can manage to come up with such a test in a reasonable time.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 6]
```diff
@@ -644,4 +649,44 @@ private List<Long> parseSupplementalGroupList(String gids) {
...
+    @Extension
+    public static class RestrictedPssSecurityContextInjector implements PodDecorator {
+
+        @NonNull
+        @Override
+        public Pod decorate(@NonNull KubernetesCloud kubernetesCloud, @NonNull Pod pod) {
+            if (kubernetesCloud.isRestrictedPssSecurityContext()) {
+                Optional<Container> maybeJNLP = pod.getSpec().getContainers().stream().filter(container -> JNLP_NAME.equals(container.getName())).findFirst();
+
+                maybeJNLP.ifPresent(jnlp -> {
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:671]: If the `jnlp` container is not defined at this point then there is a bug somewhere. At this point this code should assume so, and if the container is not defined yet then print a SEVERE error log (as written it's just silently ignoring the problem).

**Comment by jglick** [comment]
Would the controller normally have `get` permission on the agent’s namespace? If so, the admin would not need to configure a setting in `KubernetesCloud` because the controller could simply check whether [PSA is enabled](https://kubernetes.io/docs/concepts/security/pod-security-admission/#pod-security-admission-labels-for-namespaces).

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 6 above)
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: 81226b76

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java (hunk 4 above)
Comment (aneveux) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:174]: I was having a look at the tests in the project, and for now there is nothing around automated tests using `kind` right? Or did I miss something?

Comment (jglick) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:174]: (Without Kind, there would be no interesting test coverage in this plugin at all.)

**Review by amuniz — APPROVED** [review]

**Comment by aneveux** [comment]
I have been trying to setup an automated test with `kind` configuring PSA and making sure that I could run pipelines in agents in that configuration, but sadly cannot get it work in a timely fashion. 

@jglick I will be creating a follow-up task to write such an automated test. Is this a strong requirement for you for merging that PR, or is it OK for you to merge this PR as is, and receive a PR later on for adding test coverage?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 7]
```diff
@@ -303,9 +307,10 @@ public Pod build() {
 
         // default jnlp container
         Optional<Container> jnlpOpt = pod.getSpec().getContainers().stream().filter(c -> JNLP_NAME.equals(c.getName()))
-                .findFirst();
+                                         .findFirst();
         Container jnlp = jnlpOpt.orElse(new ContainerBuilder().withName(JNLP_NAME)
-                .withVolumeMounts(volumeMounts.values().toArray(new VolumeMount[volumeMounts.values().size()])).build());
+                                                              .withVolumeMounts(volumeMounts.values().toArray(new VolumeMount[volumeMounts.values().size()])).build());
+
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:313]: Can you revert unrelated whitespace changes please?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 8]
```diff
@@ -644,4 +649,46 @@ private List<Long> parseSupplementalGroupList(String gids) {
...
+                                                    .withDrop("ALL")                    //
+                                                .endCapabilities()                      //
+                                                .withRunAsNonRoot()                     //
+                                                .withNewSeccompProfile()                //
+                                                    .withType("RuntimeDefault")         //
+                                                .endSeccompProfile()                    //
+                                            .build());                                  //
+                }, () -> {
+                    LOGGER.severe(() -> "Cannot find the jnlp container to configure its security when it is expected to be present.");
+                });
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:689]: or 
```suggestion
                }, () -> LOGGER.severe("Cannot find the jnlp container to configure its security when it is expected to be present."));
```
(though this message with no context is going to be useless)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 9]
```diff
@@ -644,4 +649,46 @@ private List<Long> parseSupplementalGroupList(String gids) {
...
+
+        @NonNull
+        @Override
+        public Pod decorate(@NonNull KubernetesCloud kubernetesCloud, @NonNull Pod pod) {
+            if (kubernetesCloud.isRestrictedPssSecurityContext()) {
+                Optional<Container> maybeJNLP = pod.getSpec().getContainers().stream().filter(container -> JNLP_NAME.equals(container.getName())).findFirst();
+
+                maybeJNLP.ifPresentOrElse(jnlp -> {
+                    if (jnlp.getSecurityContext() != null) {
+                        LOGGER.warning(() -> "Overriding the existing JNLP container Security Context due to the configured restricted PSP injection");
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:673]: This is going to go into the global log and not be actionable at all.

Anyway I think this should be merging in the specified changes, rather than clobbering unrelated `securityContext` fields like specifying an FS group or whatever.

**Comment by jglick** [comment]
> Is this a strong requirement for you

I am not the plugin maintainer, but it seems optional. I can imagine such a test being tricky.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 10]
```diff
@@ -644,4 +649,46 @@ private List<Long> parseSupplementalGroupList(String gids) {
...
+     * </p>
+     */
+    @Extension
+    public static class RestrictedPssSecurityContextInjector implements PodDecorator {
+
+        @NonNull
+        @Override
+        public Pod decorate(@NonNull KubernetesCloud kubernetesCloud, @NonNull Pod pod) {
+            if (kubernetesCloud.isRestrictedPssSecurityContext()) {
+                Optional<Container> maybeJNLP = pod.getSpec().getContainers().stream().filter(container -> JNLP_NAME.equals(container.getName())).findFirst();
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:669]: Given

https://github.com/jenkinsci/kubernetes-plugin/blob/746e28c7546b7156692a162927cd52129762d8fb/src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java#L305-L311

not having a JNLP container is an error case. You should be throwing `IllegalStateException`

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+      allowPrivilegeEscalation: false
+      capabilities:
+        drop:
+        - ALL
+      runAsNonRoot: true
+      seccompProfile:
+        type: RuntimeDefault
+        </code>
+    </pre>
+</div>
```
Comment (jeromepochat) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html:26]: Remove extra spacing
```suggestion
    <pre>securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
      - ALL
  runAsNonRoot: true
  seccompProfile:
    type: RuntimeDefault</pre>
</div>
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 8 above)
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Replaced it with an `IllegalStateException` as mentioned by Vincent

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 10 above)
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:668]: :+1: Updated accordingly. thanks!

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 9 above)
Comment (aneveux) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Oh that is a nice idea indeed.

I changed the code to update the existing `securityContext` in case one is found.

**Review by jeromepochat — CHANGES_REQUESTED** [review]

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (aneveux) [reply src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html]: done in [b94a481](https://github.com/jenkinsci/kubernetes-plugin/pull/1406/commits/b94a481db31cc2882288c8c60d70345aa7479aa2)

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+securityContext:
+  allowPrivilegeEscalation: false
+  capabilities:
+    drop:
+    - ALL
+  runAsNonRoot: true
+  seccompProfile:
+    type: RuntimeDefault
+        </code>
+    </pre>
```
Comment (jeromepochat) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/help-restrictedPssSecurityContext.html:25]: Still some extra empty lines that could be removed:
```suggestion
    <pre>securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
    - ALL
  runAsNonRoot: true
  seccompProfile:
    type: RuntimeDefault</pre>
```

**Comment by aneveux** [comment]
Hello :wave: 

Considering there are 2 :heavy_check_mark: and no strong veto against this PR, I will merge it.

FTR, this PR was also manually tested using a `kind` cluster and ensuring that with a namespace configured with `PSA` this PR was providing a solution to [issues.jenkins.io/browse/JENKINS-71639](https://issues.jenkins.io/browse/JENKINS-71639).
